### DISCUSSION TITLE
Does Update Reset the TTL?

### DISCUSSION THREAD

**Discussion by ShiromMakkad** [moka-rs/moka#450] [Q&A] [answered]
Hello, I'm wondering if updating the value inside the cache will reset the TTL. I'm specifically wondering about `and_try_compute_with` and `and_upsert_with`. Also, will it update the time to idle? I'm going to guess it's yes but I want to make sure.

**Comment by ShiromMakkad** [comment]
I wrote a test and confirmed that `and_try_compute_with` does update the ttl. One other question, do I need to run https://docs.rs/moka/latest/moka/future/struct.Cache.html#method.run_pending_tasks for the ttl to update? In my test, I didn't need to, but I'm not sure if that's just luck. What tasks does that run? Size based evictions?

**Comment by tatsuya6502** [comment, marked as answer]
Hi. Yes, `and_try_compute_with` and other `insert` like methods update the TTL and time to idle timers. 

The following diagram (from #72) is a bit outdated, but `last_modified` field in the middle right box is for the TTL, and `last_accessed` field is for the time to idle.

![moka-mem-overhead-v0 7 2-2022-02-06](https://user-images.githubusercontent.com/159403/152669190-8cb43787-a274-4891-a408-217df2f09d72.png)

The "Key" box and "Value" boxes (including the `last_modified` field) are updated when an `insert` like method is called, so you do not need to call `run_pending_tasks`. The boxes on the bottom will be updated by an implicit or explicit call on `run_pending_tasks`. They are used to evict entries when the cache is full.

The tasks ran by `run_pending_tasks` are documented [here](https://github.com/moka-rs/moka/blob/main/MIGRATION-GUIDE.md#the-maintenance-tasks).

For example, it removes expired entries:

> 4. Remove expired entries.

And you would not have to call `run_pending_tasks` explicitly to remove expired entries:

> Note that cache read methods such as the `get`, `get_with` and `contains_key` never return expired entries although they are not removed immediately from the cache when they expire. You will not need to call `run_pending_tasks` method to remove expired entries unless you want to remove them immediately (e.g. to free some resources).

(I should put the info to a better place in the document, but I have not found the time to do so yet.)

**Reply by ShiromMakkad** [reply]
This answers everything perfectly, thanks! I really appreciate your creation, support and maintenance of this repo; it's really well done.
